functions/go/set-namespace/main.go

```diff
@@ -1,75 +1,139 @@
-		clusterScopedResources := []string{}
-		for i := range resourceList.Items {
-			// modify the resources using the kyaml/yaml library:
-			// https://pkg.go.dev/sigs.k8s.io/kustomize/kyaml/yaml
-			node := resourceList.Items[i]
-			kindNode, err := node.Pipe(yaml.Lookup("kind"))
-			if err != nil {
-				// ignore the node if it does not have a "kind" field
-				continue
-			}
-			kind := yaml.GetValue(kindNode)
-			typeMeta := yaml.TypeMeta{
-				Kind: kind,
-			}
-			if typeMeta.IsNamespaceable() {
-				// Set the metadata.namespace field
-				if err := node.PipeE(yaml.LookupCreate(
-					yaml.ScalarNode, "metadata", "namespace"),
-					yaml.FieldSetter{StringValue: namespace}); err != nil {
-					return err
-				}
-			} else {
-				if !allowClusterScoped {
-					apiVersionNode, err := node.Pipe(yaml.Lookup("apiVersion"))
-					if err != nil {
-						continue
-					}
-					apiVersion := yaml.GetValue(apiVersionNode)
-					nameNode, err := node.Pipe(yaml.Lookup("metadata", "name"))
-					if err != nil {
-						continue
-					}
-					name := yaml.GetValue(nameNode)
-					resID := fmt.Sprintf("apiVersion: %s, kind: %s, name: %s", apiVersion, kind, name)
-					clusterScopedResources = append(clusterScopedResources, resID)
-				}
-			}
+		resMap, err := resmapFactory.NewResMapFromRNodeSlice(resourceList.Items)
+		if err != nil {
+			return err
 		}
-		if len(clusterScopedResources) > 0 {
-			return fmt.Errorf("the app config should only include namespace-scoped resources. "+
-				"But the following cluster-scoped resources are found:\n%s\n", strings.Join(clusterScopedResources, "\n"))
+		dataField, err := getDataFromFunctionConfig(resourceList.FunctionConfig)
+		if err != nil {
+			return err
+		}
+		dataValue, err := yaml.Marshal(dataField)
+		if err != nil {
+			return err
 		}
-		return nil
-	})
 
-	cmd.Flags().BoolVar(&allowClusterScoped, "allow-cluster-scoped", true, "allow cluster-scoped resources or not. "+
-		"This function may not be able to identify all the cluster-scoped resources "+warning)
-	cmd.Flags().StringVar(&namespace, "namespace", "", "the namespace to be added into namespaced resources "+
-		"This function may set the metadata.namespace field of some cluster-scoped resources "+warning)
+		err = plugin.Config(pluginHelpers, dataValue)
+		if err != nil {
+			return err
+		}
+		if len(plugin.FieldSpecs) == 0 {
+			plugin.FieldSpecs = defaultConfig
+		}
+		err = plugin.Transform(resMap)
+		if err != nil {
+			return err
+		}
 
+		resourceList.Items, err = resMap.ToRNodeSlice()
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	cmd.Long = usage()
 	if err := cmd.Execute(); err != nil {
+		fmt.Println(err)
 		os.Exit(1)
 	}
 }
+
+func usage() string {
+	return `Update or add namespace.
+
+Configured using a ConfigMap with the following keys:
+
+namespace: Name of the namespace that will be set.
+
+Example:
+
+To add a namespace 'color' to all resources:
+
+apiVersion: v1
+kind: ConfigMap
+data:
+  namespace: color
+metadata:
```

Comment:
by convention, 'metadata' field should come before 'data'.

Reply:
This metadata is for function config. Should it be from data as wel?

Reply:
Yes, this is general convention for any KRM that GVKNN should come before the spec.